pkg/util/policy/agent/agent.py

```diff
@@ -284,138 +258,231 @@ def make_new_full_ruleset(current_rules, new_rules):
+            # tid/sid - detected by tenant_network_id and segment_network_id
+            pr = peer.get('peer')
+            cidr = peer.get('cidr')
+            from_tenant = peer.get('tenant_network_id')
+            from_segment = peer.get('segment_network_id')
+
+            if pr:
+                if pr == "any":
+                    jump_rules = [ _make_rule(policy_chain_name, "-j %s") % in_chain_name ]
+                else:
+                    raise Exception("Unsupported value of peer %s" % pr)
+
+            elif cidr:
+                jump_rules = [ _make_rule(policy_chain_name, "-s %s -j %s") % (cidr, in_chain_name) ]
+
+            elif from_segment and from_tenant:
+                u32_in_match = _make_u32_match(addr_scheme, tenant, from_segment)
+                jump_rules = [
+                    _make_rule(policy_chain_name, '-m u32 --u32 "%s" -j %s' %
+                                                       (u32_in_match, in_chain_name))
+                ]
+
+            else:
+                raise Exception("Unknown peer type %s" % pr)
+
+            if not policy_chain_name in rules:
+                rules[policy_chain_name] = []
+                rules[policy_chain_name].append(
+                    _make_rule(policy_chain_name, '-m comment --comment PolicyId=%s -j RETURN' % policy_id))
+
+            for r in jump_rules:
+                rules[policy_chain_name].insert(0,r)
+
+    # Loop over rules and render protocol + action part of each rules
+    for r in _make_rules(policy_def.get('rules')):
+        for pc in policy_chains:
+            if not pc in rules:
+                rules[pc] = []
+        # and stuff copy of the rule into each policy chain
+        rules[pc].append(_make_rule(pc,r))
 
     return rules
 
+
+def _make_rules(policy_rules):
+    """
+    For each rules in policy_rules creates in/out rules in iptables_rules.
+
+    Returns updated list of iptables_rules
+    """
+    # For each rule in the policy create iptables rules.
+    for r in policy_rules:
+        if r['protocol'] == 'TCP':
+            stateful = r.get("is_stateful")
+            if stateful:
+                raise Exception("Flag is_stateful not implemented")
+            else:
+                in_rules = [ '-p tcp --dport %s -j ACCEPT' % ':'.join(str(x) for x in r["ports"]) ]
+
+        elif r['protocol'] == 'UDP':
+            in_rules =[ '-p udp --dport %s -j ACCEPT' % ':'.join(str(x) for x in r["ports"]) ]
+
+        elif r['protocol'] == 'ICMP':
+            icmp_types = r.get('icmp_type_code')
+            in_rules = []
+            if icmp_types:
+                for icmp_type in icmp_types:
+                    in_rules.append('-p icmp --icmp-type %s -j ACCEPT' % icmp_type)
+            in_rules.append('-p icmp -j ACCEPT')
+
+        elif r['protocol'] == 'any':
+            in_rules = [ '-j ACCEPT' ]
+
+        else:
+            in_rules = [ '-m comment --comment error_unknownProtocol -j LOG' ]
```

Comment:
Good question, do we want to apply other rules in a list if one of them broken, or blow up loud ?